targets/TARGET_Analog_Devices/TARGET_ADUCM4X50/TARGET_ADUCM4050/api/serial_api.c

```diff
@@ -0,0 +1,251 @@
+ *   - This software may not be combined or merged with other code in any manner
+ *     that would cause the software to become subject to terms and conditions
+ *     which differ from those listed here.
+ *   - Neither the name of Analog Devices, Inc. nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *   - The use of this software may or may not infringe the patent rights of one
+ *     or more patent holders.  This license does not release you from the
+ *     requirement that you obtain separate licenses from these patent holders
+ *     to use this software.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES, INC. AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-
+ * INFRINGEMENT, TITLE, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL ANALOG DEVICES, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, PUNITIVE OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, DAMAGES ARISING OUT OF
+ * CLAIMS OF INTELLECTUAL PROPERTY RIGHTS INFRINGEMENT; PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ ******************************************************************************/
+
+#include "serial_api.h"
+
+#if DEVICE_SERIAL
+
+// math.h required for floating point operations for baud rate calculation
+#include <math.h>
+#include "mbed_assert.h"
+
+#include <string.h>
+
+#include "cmsis.h"
+#include "pinmap.h"
+#include "PeripheralPins.h"
+#include "drivers/uart/adi_uart.h"
+#define ADI_UART_MEMORY_SIZE   (ADI_UART_BIDIR_MEMORY_SIZE)
+#define ADI_UART_NUM_DEVICES   2
+
+static ADI_UART_HANDLE hDevice[ADI_UART_NUM_DEVICES];
+static uint32_t UartDeviceMem[ADI_UART_NUM_DEVICES][(ADI_UART_MEMORY_SIZE + 3)/4];
+static uint32_t serial_irq_ids[ADI_UART_NUM_DEVICES] = {0};
+static uart_irq_handler irq_handler = NULL;
+int stdio_uart_inited = 0;
+serial_t stdio_uart;
+static int rxbuffer[2];
+static int txbuffer[2];
+
+static void uart_callback(void *pCBParam, uint32_t Event, void *pArg)
+{
+    MBED_ASSERT(irq_handler);
+    serial_t *obj = pCBParam;
+    if (Event == ADI_UART_EVENT_TX_BUFFER_PROCESSED)
+        irq_handler(serial_irq_ids[obj->index], TxIrq);
+    else if (Event == ADI_UART_EVENT_RX_BUFFER_PROCESSED)
+        irq_handler(serial_irq_ids[obj->index], RxIrq);
+}
+
+
+void serial_free(serial_t *obj)
+{
+    adi_uart_Close(hDevice[obj->index]);
+}
+
+void serial_baud(serial_t *obj, int baudrate)
+{
+    uint32_t uartdivc,uartdivm,uartdivn,uartosr;
+
+    // figures based on PCLK of 26MHz
```

Comment:
How does this work if PCLK is changed? Shouldn't t his get a value of PCLK and recalculate ?

Reply:
The baud rates also depend on divn, divm etc., which are different for a different PCLK rate, so there is no simple formula to vary these values based on PCLK. One thing we can do is to have a look up table that have these pre-set values for a couple of different PCLK rates that users typically will use.